## Business understanding
The idea behind association rule mining is to find patterns between different variables. The goal of this assignment is to find relationships between purchases across multiple product groups.

## Data understanding
The data used contains 20 different product groups with column values of 1 or 0, where 1 indicates that a product was purchased from that group and 0 indicates that a purchase didn't include a product from that group. There are no missing values in the dataset.

## Data preparation
The data needed minimal preparation. The 'ID' column was dropped as it isn't needed and all the values in the columns were transformed into boolean values, where 1 = True and 0 = False, because the mlxtend library used requires the usage of boolean values. 

In [1]:
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
df = pd.read_csv('data/drone_prod_groups.csv', sep=',')

# Data preprocessing
df = df.drop(columns='ID')
df = df.apply(lambda x: x.map(lambda y: True if y == 1 else False))
df.head()

Prod1  Prod2  Prod3  Prod4  Prod5  Prod6  Prod7  Prod8  Prod9  Prod10  \
0  False  False  False  False  False  False  False  False   True   False   
1  False   True  False  False  False  False  False  False   True   False   
2  False  False  False  False  False  False   True  False  False   False   
3   True  False  False   True  False  False  False  False  False   False   
4  False  False  False  False  False  False  False  False   True   False   

   Prod11  Prod12  Prod13  Prod14  Prod15  Prod16  Prod17  Prod18  Prod19  \
0   False   False   False   False    True   False   False   False   False   
1   False   False   False   False    True    True    True    True    True   
2   False   False   False   False   False   False   False   False    True   
3    True   False   False   False   False   False   False   False    True   
4   False   False   False   False    True   False   False   False    True   

   Prod20  
0    True  
1    True  
2    True  
3    True  
4    True

## Modeling
To find the relationships in the data we used the Apriori algorithm with the implementation of the algorithm imported from mlxtend library. The Apriori algorithm is a popular algorithm for mining frequent itemsets for boolean association rules. The algorithm is designed to operate on databases containing transactions. The algorithm finds all frequent item sets and uses a candidate generation function to generate all possible itemsets. The algorithm terminates when no more itemsets can be found. \
We only used one hyperparameter: min_support. Support is the proportion of transactions in the database that contain the item set, in this case an item set is a set of product groups present in a transaction. In our case we decided on 3% of transactions have to include the itemset for it to be significant for our analysis.

In [2]:
freq = apriori(df, min_support=0.03, use_colnames=True)
# Filters the itemsets with one product group and breaks the next code cell
# freq= freq[freq['itemsets'].apply(lambda x: len(x) != 1)] 
freq

support                    itemsets
0   0.10998                    ( Prod1)
1   0.13098                    ( Prod2)
2   0.03271                    ( Prod3)
3   0.03585                    ( Prod4)
4   0.10459                    ( Prod5)
5   0.13499                    ( Prod7)
6   0.16179                    ( Prod8)
7   0.19853                    ( Prod9)
8   0.09336                   ( Prod10)
9   0.10848                   ( Prod11)
10  0.15971                   ( Prod12)
11  0.14557                   ( Prod14)
12  0.11880                   ( Prod15)
13  0.13100                   ( Prod16)
14  0.05618                   ( Prod17)
15  0.12166                   ( Prod18)
16  0.20626                   ( Prod19)
17  0.14798                   ( Prod20)
18  0.03210            ( Prod9,  Prod2)
19  0.03346           ( Prod19,  Prod2)
20  0.06683           ( Prod5,  Prod12)
21  0.03334            ( Prod7,  Prod9)
22  0.03411           ( Prod7,  Prod19)
23  0.03893            ( Prod9,  Prod8)
24  0.03164           ( Prod12,  Prod8)
25  0.04011           ( Prod19,  Prod8)
26  0.03845           ( Prod9,  Prod12)
27  0.03484           ( Prod14,  Prod9)
28  0.11145           ( Prod15,  Prod9)
29  0.03160           ( Prod16,  Prod9)
30  0.04996           ( Prod9,  Prod19)
31  0.03676           ( Prod9,  Prod20)
32  0.03881          ( Prod12,  Prod19)
33  0.03607          ( Prod14,  Prod19)
34  0.03041          ( Prod15,  Prod19)
35  0.03296          ( Prod16,  Prod19)
36  0.03063          ( Prod18,  Prod19)
37  0.13476          ( Prod20,  Prod19)
38  0.03345  ( Prod20,  Prod9,  Prod19)

In [3]:
rules = association_rules(freq, metric='lift', min_threshold=0.5).sort_values(by='lift', ascending=False)
rules

antecedents         consequents  antecedent support  \
21            ( Prod9)           ( Prod15)             0.19853   
20           ( Prod15)            ( Prod9)             0.11880   
42   ( Prod9,  Prod19)           ( Prod20)             0.04996   
43           ( Prod20)   ( Prod9,  Prod19)             0.14798   
39           ( Prod19)           ( Prod20)             0.20626   
38           ( Prod20)           ( Prod19)             0.14798   
40   ( Prod9,  Prod20)           ( Prod19)             0.03676   
45           ( Prod19)   ( Prod9,  Prod20)             0.20626   
4             ( Prod5)           ( Prod12)             0.10459   
5            ( Prod12)            ( Prod5)             0.15971   
27           ( Prod20)            ( Prod9)             0.14798   
26            ( Prod9)           ( Prod20)             0.19853   
44            ( Prod9)  ( Prod20,  Prod19)             0.19853   
41  ( Prod20,  Prod19)            ( Prod9)             0.13476   
6             ( Prod7)            ( Prod9)             0.13499   
7             ( Prod9)            ( Prod7)             0.19853   
33           ( Prod19)           ( Prod15)             0.20626   
32           ( Prod15)           ( Prod19)             0.11880   
2            ( Prod19)            ( Prod2)             0.20626   
3             ( Prod2)           ( Prod19)             0.13098   
1             ( Prod2)            ( Prod9)             0.13098   
0             ( Prod9)            ( Prod2)             0.19853   
9            ( Prod19)            ( Prod7)             0.20626   
8             ( Prod7)           ( Prod19)             0.13499   
12           ( Prod12)            ( Prod8)             0.15971   
13            ( Prod8)           ( Prod12)             0.16179   
36           ( Prod18)           ( Prod19)             0.12166   
37           ( Prod19)           ( Prod18)             0.20626   
25           ( Prod19)            ( Prod9)             0.20626   
24            ( Prod9)           ( Prod19)             0.19853   
34           ( Prod16)           ( Prod19)             0.13100   
35           ( Prod19)           ( Prod16)             0.20626   
22           ( Prod16)            ( Prod9)             0.13100   
23            ( Prod9)           ( Prod16)             0.19853   
17           ( Prod12)            ( Prod9)             0.15971   
16            ( Prod9)           ( Prod12)             0.19853   
10            ( Prod9)            ( Prod8)             0.19853   
11            ( Prod8)            ( Prod9)             0.16179   
19            ( Prod9)           ( Prod14)             0.19853   
18           ( Prod14)            ( Prod9)             0.14557   
14           ( Prod19)            ( Prod8)             0.20626   
15            ( Prod8)           ( Prod19)             0.16179   
31           ( Prod19)           ( Prod14)             0.20626   
30           ( Prod14)           ( Prod19)             0.14557   
29           ( Prod19)           ( Prod12)             0.20626   
28           ( Prod12)           ( Prod19)             0.15971   

    consequent support  support  confidence      lift  representativity  \
21             0.11880  0.11145    0.561376  4.725388               1.0   
20             0.19853  0.11145    0.938131  4.725388               1.0   
42             0.14798  0.03345    0.669536  4.524501               1.0   
43             0.04996  0.03345    0.226044  4.524501               1.0   
39             0.14798  0.13476    0.653350  4.415125               1.0   
38             0.20626  0.13476    0.910664  4.415125               1.0   
40             0.20626  0.03345    0.909956  4.411696               1.0   
45             0.03676  0.03345    0.162174  4.411696               1.0   
4              0.15971  0.06683    0.638971  4.000822               1.0   
5              0.10459  0.06683    0.418446  4.000822               1.0   
27             0.19853  0.03676    0.248412  1.251256               1.0   
26      

## Evaluation
The algorithm found 39, 21 if we exclude the single item itemsets, different itemsets where the product groups were present in at least 3% of the transactions in the dataset. \
For association rules we used the 'Lift' metric. Lift is a derived metric that tells the factor with which the probability of the consequent being present increases when the antecedent is present i.e. buying a product from group 9 increases the probability of buying a product from group 15 by a factor of 4.7.

## Deployment
The model can be used in anything involving commerce. Different association rules should be utilized when deciding how to place products in a store or what to recommend on an e-commrece platform e.g. an online shop selling electronics could look at the lift metric and use that to determine what other products to recommend when a customer adds a product to their checkout basket. A physical store could use the support metric to see which items are often purchased together and place the items in the store accordingly, either far apart from each other if they want to keep the customer browsing and impulse buying for as long as possible or close to each other for convenience.